enumerateDevices does not need to check for device permission

youennf opened the pull request:
#### 5d898848eda91953c7e801cc0d356f77004faa5b
<pre>
enumerateDevices does not need to check for device permission
<a href="https://rdar.apple.com/148094614">rdar://148094614</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290619">https://bugs.webkit.org/show_bug.cgi?id=290619</a>

Reviewed by Eric Carlson.

Align with the spec and Firefox by not taking into account permissions in the enumeration of devices.
This improves privacy in the case a user decided to grant persistently camera.
We add a runtime flag for safety.

We also add a quirk for Google Meet where we replace deviceId/label from empty string to &quot;default&quot;.
And similary in getUserMedia, we remove exact deviceId constraint with &quot;default&quot; value.

Tested on existing websites and covered by added and rebased tests.

* LayoutTests/fast/mediastream/enumerateDevices-camera-denied-expose-with-capture-expected.txt: Added.
* LayoutTests/fast/mediastream/enumerateDevices-camera-denied-expose-with-capture.html: Copied from LayoutTests/fast/mediastream/enumerateDevices-camera-denied.html.
* LayoutTests/fast/mediastream/enumerateDevices-camera-denied.html:
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied-expose-with-capture-expected.txt: Added.
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied-expose-with-capture.html: Copied from LayoutTests/fast/mediastream/enumerateDevices-microphone-denied.html.
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied.html:
* LayoutTests/fast/mediastream/media-devices-enumerate-devices.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices-persistent-permission.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices.https-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::shouldEnableEnumerateDeviceQuirk):
(WebCore::createMediaConstraints):
(WebCore::MediaDevices::getUserMedia):
(WebCore::MediaDevices::exposeDevices):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableEnumerateDeviceQuirk const):
(WebCore::handleGoogleQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):

Canonical link: <a href="https://commits.webkit.org/293003@main">https://commits.webkit.org/293003@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/844c216a31c4c21b00e4997be5bf5dd2943eba29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7347 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102595 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48035 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99551 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17424 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25583 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74304 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31485 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88182 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54649 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6063 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47477 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90182 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6144 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104613 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96128 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24585 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17937 "Found 2 new test failures: imported/w3c/web-platform-tests/workers/data-url-shared.html ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83354 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84307 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82775 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20872 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27313 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5001 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18179 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24547 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29716 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119754 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24369 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33612 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27683 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25943 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -1033,23 +1045,11 @@ void UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame(Frame
     ALWAYS_LOG(LOGIDENTIFIER);
 
-    auto callback = [this, protectedThis = Ref { *this }, frameID, userMediaDocumentOrigin, topLevelDocumentOrigin, completionHandler = WTFMove(completionHandler)](PermissionInfo permissionInfo) mutable {
+    auto callback = [this, protectedThis = Ref { *this }, frameID, userMediaDocumentOrigin, topLevelDocumentOrigin, completionHandler = WTFMove(completionHandler)](auto cameraState, auto microphoneState) mutable {
```
Can we use a WeakPtr here instead of a Ref?

youennf replied:
Yep, will do

eric-carlson commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2293,6 +2298,9 @@ static void handleGoogleQuirks(QuirksData& quirksData, const URL& quirksURL, con
     quirksData.isTouchBarUpdateSuppressedForHiddenContentEditableQuirk = quirksData.isGoogleDocs;
 #endif
+#if ENABLE(MEDIA_STREAM)
+    quirksData.shouldEnableEnumerateDeviceQuirk = true;
```
Do we need to enable this for all Google properties?

youennf replied:
I'll change to Google Meet for now

eric-carlson commented on Source/WebCore/Modules/mediastream/MediaDevices.cpp:
```diff
@@ -105,7 +110,15 @@ static MediaConstraints createMediaConstraints(const std::variant<bool, MediaTra
-            return createMediaConstraints(trackConstraints);
+            auto result = createMediaConstraints(trackConstraints);
+            if (document && shouldEnableEnumerateDeviceQuirk(*document)) {
+                if (result.isValid && result.mandatoryConstraints.deviceId() && result.mandatoryConstraints.deviceId()->isMandatory()) {
```
Not: this would be slightly easier to read if there was a local with `result.mandatoryConstraints.deviceId()`

youennf replied:
Will change

youennf committed:
enumerateDevices does not need to check for device permission
rdar://148094614
https://bugs.webkit.org/show_bug.cgi?id=290619

Reviewed by Eric Carlson.

Align with the spec and Firefox by not taking into account permissions in the enumeration of devices.
This improves privacy in the case a user decided to grant persistently camera.
We add a runtime flag for safety.

We also add a quirk for Google Meet where we replace deviceId/label from empty string to "default".
And similary in getUserMedia, we remove exact deviceId constraint with "default" value.

Tested on existing websites and covered by added and rebased tests.

* LayoutTests/fast/mediastream/enumerateDevices-camera-denied-expose-with-capture-expected.txt: Added.
* LayoutTests/fast/mediastream/enumerateDevices-camera-denied-expose-with-capture.html: Copied from LayoutTests/fast/mediastream/enumerateDevices-camera-denied.html.
* LayoutTests/fast/mediastream/enumerateDevices-camera-denied.html:
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied-expose-with-capture-expected.txt: Added.
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied-expose-with-capture.html: Copied from LayoutTests/fast/mediastream/enumerateDevices-microphone-denied.html.
* LayoutTests/fast/mediastream/enumerateDevices-microphone-denied.html:
* LayoutTests/fast/mediastream/media-devices-enumerate-devices.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices-persistent-permission.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices.https-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::shouldEnableEnumerateDeviceQuirk):
(WebCore::createMediaConstraints):
(WebCore::MediaDevices::getUserMedia):
(WebCore::MediaDevices::exposeDevices):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableEnumerateDeviceQuirk const):
(WebCore::handleGoogleQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):

Canonical link: https://commits.webkit.org/293003@main
